kernel/src/hil/symmetric_encryption.rs

```diff
@@ -1,42 +1,87 @@
-    fn aes128_crypt_ctr(&self, data: &'static mut [u8], init_ctr: &'static mut [u8], len: usize);
+    /// Set the IV (or initial counter); returns `EINVAL` if length is not `AES128_BLOCK_SIZE`
+    fn set_iv(&'a self, iv: &'a [u8]) -> ReturnCode;
+
+    /// Set the source buffer.  If this is full, the encryption
+    /// input will be this entire buffer, and its size must match
+    /// `stop_index - start_index` when `crypt()` is called.
+    /// If this is empty, the destination buffer will be read
+    /// to provide the plaintext input.
+    fn set_source(&'a self, buf: Option<&'a [u8]>) -> ReturnCode;
+
+    /// Set the destination buffer.  If `set_source()` has not
+    /// been used to pass a source buffer, this buffer will also
+    /// provide the encryption input, which will be overwritten.
+    /// The option should be full whenever `crypt()` is called.
+    /// Returns SUCCESS if the buffer was installed, or EBUSY
+    /// if the encryption unit is still busy.
+    fn put_dest(&'a self, dest: Option<&'a mut [u8]>) -> ReturnCode;
+
+    /// Return the destination buffer, if any.
+    /// Returns EBUSY if the encryption unit is still busy.
+    fn take_dest(&'a self) -> Result<Option<&'a mut [u8]>, ReturnCode>;
+
+    /// Begin a new message (with the configured IV) when `crypt()` is next
+    /// called.  Multiple calls to `put_data()` and `crypt()` may be made
+    /// between calls to `start_message()`, allowing the encryption context
+    /// to extend over non-contiguous extents of data.
+    fn start_message(&self);
+
+    /// Request an encryption/decryption
+    ///
+    /// The indices `start_index` and `stop_index` must be valid offsets in
+    /// a buffer previously passed in with `set_data`, and the length
+    /// `stop_index - start_index` must be a multiple of
+    /// `AES128_BLOCK_SIZE`.  Otherwise, INVAL will be returned.
+    ///
+    /// If SUCCESS is returned, the client's `crypt_done` method will eventually
+    /// be called, and the portion of the data buffer between `start_index`
+    /// and `stop_index` will hold the result of the encryption/decryption.
+    ///
+    /// For correct operation, the methods `set_key` and `set_iv` must have
+    /// previously been called to set the buffers containing the
+    /// key and the IV (or initial counter value), and a method `set_mode_*()`
+    /// must have been called to set the desired mode.  These settings persist
+    /// across calls to `crypt()`.
+    fn crypt(&self,
+             start_index: usize,
+             stop_index: usize) -> ReturnCode;
 }
 
-pub trait Client {
-    /// send back to result of the encryption/decryption to the capsule
-    /// this should be hardware independent if the cryptostate is used for all
-    /// implementations
-    fn crypt_done(&self,
-                  data: &'static mut [u8],
-                  dmy: &'static mut [u8],
-                  len: usize)
-                  -> ReturnCode;
+pub trait AES128Ctr {
```

Comment:
Just looking towards eventually maybe exposing this to userland, this method of switching modes probably means that the it doesn't make sense for the userland driver be generic across implementations that have different modes. Or at least I can't see how we would do that. That's probably fine, but, at least for cross-platform code, it means we won't use something like Java's crypto interface where you pass some string requesting some algorithm, and the library might throw up in your face.

Reply:
To be clear, I'm not asking for a change here, Just raising the issue.